urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/controllers/momcts/utils/SetOperation.java

```diff
@@ -0,0 +1,673 @@
+	 * @param maxseq
+	 * @return
+	 */
+	public static <T> Vector<T> joinSeq(Vector<T> s, Vector<T> segment, int maxseq){
+		@SuppressWarnings("unchecked")
+		Vector<T> seqOrg = (Vector<T>) s.clone();
+		int size = segment.size();
+
+		if(maxseq < 0 || seqOrg.size()+ size < maxseq){
+			seqOrg = joinSeq(seqOrg, segment);
+		} else {
+			int offset = Math.min(size+seqOrg.size()-maxseq,size);
+			eliminateFirstNElements(seqOrg, offset);
+			seqOrg = joinSeq(seqOrg, segment);
+		}
+		
+		return seqOrg;
+	}
+	
+	public static <T> Vector<T> joinSeq(T v, Vector<T> v2){
+		Vector<T> v1 = new Vector<T>();
+		v1.add(v);
+		return joinSeq(v1,v2);
+	}
+	
+	public static <T> Vector<T> joinSeq(T v, Vector<T> v2, int maxseq){
+		Vector<T> v1 = new Vector<T>();
+		v1.add(v);
+		return joinSeq(v1,v2, maxseq);
+	}
+	
+	public static <T> Vector<T> joinSeq(Vector<T> v1, T v){
+		Vector<T> v2 = new Vector<T>();
+		v2.add(v);
+		return joinSeq(v1,v2);
+	}
+	
+	public static <T> Vector<T> joinSeq(Vector<T> v1, T v, int maxseq){
+		Vector<T> v2 = new Vector<T>();
+		v2.add(v);
+		return joinSeq(v1,v2, maxseq);
+	}
+	
+	/**
+	 * Count the frequency of elements in a list
+	 * @param v
+	 * @return
+	 */
+	public static <T> HashMap<T, Integer> frequence(Vector<T> v){
+		HashMap<T, Integer> freq = new HashMap<T, Integer>();
+		for(T el : v){
+			if(freq.containsKey(el)) freq.put(el, freq.get(el)+1);
+			else freq.put(el, 1);
+		}
+		return freq;
+	}
+	
+	public static <T> HashMap<T, Double> frequencePortion(Vector<T> v){
+		HashMap<T, Double> freq = new HashMap<T, Double>();
+		for(T el : v){
+			if(freq.containsKey(el)) freq.put(el, freq.get(el)+1);
+			else freq.put(el, 1.0);
+		}
+		return divs(freq,v.size());
+	}
+	
+	public static <T> Vector<T> ones(int length, T initV){
+		Vector<T> os = new Vector<T>();
+		for(int i=0; i<length; i++){
+			os.add(initV);
+		}
+		return os;
+	}
+	
+	public static Vector<Integer> arithmeticProgression(int length, int initV, int diff){
+		Vector<Integer> progs = new Vector<Integer>();
+		for(int i=0; i<length; i++){
+			progs.add(initV);
+			initV += diff;
+		}
+		return progs;
+	}
+	
+	public static Vector<Integer> arithmeticProgression(int length, int initV){
+		return arithmeticProgression(length, initV, 1);
+	}
+	
+	public static Vector<Integer> arithmeticProgression(int length){
+		return arithmeticProgression(length, 0);
+	}
+	
+	public static Vector<Double> arithmeticProgression(int length, double initV, double diff){
+		Vector<Double> progs = new Vector<Double>();
+		for(int i=0; i<length; i++){
+			progs.add(initV+diff);
+			initV += diff;
+		}
+		return progs;
+	}
+	
+	/**
+	 * Return a vector containing only one indicated element
+	 * @return
+	 */
+	public static <T> Vector<T> singleton(T el){
+		Vector<T> v = new Vector<T>();
+		v.add(el);
+		return v;
+	}
+	
+	public static double weightedSum(Vector<Double> v, Vector<Double> wts){
+		double sum = 0;
+		for(int i=0; i<v.size();i++){
+			sum += v.get(i)*wts.get(i%wts.size());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;wts&#39; (index: 1 | type: Vector). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;v&#39;. The same parameter type is validated here for example: urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/controllers/momcts/MOOTools.java:29. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).